proxy/src/main/java/com/velocitypowered/proxy/protocol/ProtocolUtils.java

```diff
@@ -153,6 +162,80 @@ public static void writeUuid(ByteBuf buf, UUID uuid) {
     buf.writeLong(uuid.getLeastSignificantBits());
   }
 
+  /**
+   * Reads an UUID stored as an Integer Array from the {@code buf}.
+   * @param buf the buffer to read from
+   * @return the UUID from the buffer
+   */
+  public static UUID readUuidIntArray(ByteBuf buf) {
+    long msbHigh = (long) buf.readInt() << 32;
+    long msbLow = (long) buf.readInt() & 0xFFFFFFFFL;
+    long msb = msbHigh | msbLow;
+    long lsbHigh = (long) buf.readInt() << 32;
+    long lsbLow = (long) buf.readInt() & 0xFFFFFFFFL;
+    long lsb = lsbHigh | lsbLow;
+    return new UUID(msb, lsb);
+  }
+
+  /**
+   * Writes an UUID as an Integer Array to the {@code buf}.
+   * @param buf the buffer to write to
+   * @param uuid the UUID to write
+   */
+  public static void writeUuidIntArray(ByteBuf buf, UUID uuid) {
+    buf.writeInt((int) (uuid.getMostSignificantBits() >> 32));
+    buf.writeInt((int) uuid.getMostSignificantBits());
+    buf.writeInt((int) (uuid.getLeastSignificantBits() >> 32));
+    buf.writeInt((int) uuid.getLeastSignificantBits());
+  }
+
+  /**
+   * Reads a {@link net.kyori.nbt.CompoundTag} from the {@code buf}.
+   * @param buf the buffer to read from
+   * @return {@link net.kyori.nbt.CompoundTag} the CompoundTag from the buffer
+   */
+  public static CompoundTag readCompoundTag(ByteBuf buf) {
+    int indexBefore = buf.readerIndex();
+    byte startType = buf.readByte();
+    if (startType == 0) {
+      throw new DecoderException("Invalid NBT start-type (end/empty)");
+    }
+    buf.readerIndex(indexBefore);
```

Comment:
Imo a mark + reset is cleaner than storing the offset. https://netty.io/4.0/api/io/netty/buffer/ByteBuf.html#markReaderIndex--

Reply:
mark/reset is gone in the next major release of Netty: https://github.com/netty/netty/issues/8535